src/catalog.h

```diff
@@ -644,58 +663,61 @@ typedef enum CacheType
 	_MAX_CACHE_TYPES
 } CacheType;
 
-typedef struct Catalog
+typedef enum CatalogType CatalogType;
+
+typedef struct CatalogTableInfo
+{
+	const char *schema_name;
+	const char *name;
+	Oid			id;
+	Oid			serial_relid;
+	Oid			index_ids[_MAX_TABLE_INDEXES];
+} CatalogTableInfo;
+
+typedef struct CatalogPrefix
 {
 	char		database_name[NAMEDATALEN];
 	Oid			database_id;
 	Oid			schema_id;
-	struct
-	{
-		const char *schema_name;
-		const char *name;
-		Oid			id;
-		Oid			serial_relid;
-		Oid			index_ids[_MAX_TABLE_INDEXES];
-	}			tables[_MAX_CATALOG_TABLES];
+	Oid			owner_uid;
+} CatalogPrefix;
+
+typedef struct Catalog
+{
+	CatalogPrefix prefix;
+	CatalogTableInfo tables[_MAX_CATALOG_TABLES];
 
 	Oid			cache_schema_id;
 	struct
 	{
 		Oid			inval_proxy_id;
 	}			caches[_MAX_CACHE_TYPES];
 
-	Oid			owner_uid;
 	Oid			internal_schema_id;
 	struct
 	{
 		Oid			function_id;
 	}			functions[_MAX_INTERNAL_FUNCTIONS];
 } Catalog;
 
-
 typedef struct CatalogSecurityContext
 {
 	Oid			saved_uid;
 	int			saved_security_context;
 } CatalogSecurityContext;
 
-bool		catalog_is_valid(Catalog *catalog);
+void		catalog_populate_prefix(CatalogPrefix *catalog, CatalogTableInfo *tables, int max_table, const TableInfoDef *table_ary, const TableIndexDef *index_ary, const char **serial_id_ary);
+
 Catalog    *catalog_get(void);
 void		catalog_reset(void);
 
+/* Functions should operate on a passed-in Catalog struct */
+int64		catalog_table_next_seq_id(Catalog *catalog, CatalogTable table);
 Oid			catalog_get_cache_proxy_id(Catalog *catalog, CacheType type);
 
-Oid			catalog_get_internal_function_id(Catalog *catalog, InternalFunction func);
-
-bool		catalog_become_owner(Catalog *catalog, CatalogSecurityContext *sec_ctx);
+/* Functions that modify the actual catalog table on disk */
+bool		catalog_become_owner(CatalogPrefix *catalog, CatalogSecurityContext *sec_ctx);
```

Comment:
I would actually maybe call this `catalog_base_become_owner` and then have a `catalog_become_owner` (static inline?) that just calls that and takes a catalog.

Reply:
changed to `catalog_base_become_owner`, but didn't inline because looks like a longer function.

Reply:
It probably won't matter, IIRC largely clang ignores `inline` anyway